Render fields with cardinality "MANY"

philandstuff opened the pull request:
The register register has a "fields" field which is a list of fields. This PR adds the ability to render such fields with a cardinality of MANY, across all representations.

Things of note:
- FieldValue.value() has been renamed FieldValue.getValue().  This is
  largely for EntryConverterTest, which can use the
  samePropertyValuesAs() matcher to compare LinkValue() types
- we use semicolons to separate values in CSV and TSV representations;
  this is slightly iffy. The idea was copied over from the alpha.

Things not done:
- I wanted to make FieldValue.isLink() and FieldValue.isList() into
  default methods, but there's a bug in OGNL (used by thymeleaf) which
  hides default methods from the templates.  This is fixed in OGNL
  3.0.12 (we're currently on 3.0.8).

See individual commits for more context.

philandstuff committed:
render fields with cardinality "many" in JSON

This commit ensures that we render fields with a cardinality of "many"
appropriately, in the JSON format.  It doesn't do anything for other
represenations.

philandstuff committed:
update fields.yaml from source

Copied and pasted from the field register.

philandstuff committed:
Render lists for all representations

This adds the ability to render lists to all representations.  To
capture this, I've changed the RepresentationsTest to cover a register
with lists in it.

This is a bit of a jumbo commit, sorry!

Things of note:

  - FieldValue.value() has been renamed FieldValue.getValue().  This is
    largely for EntryConverterTest, which can use the
    samePropertyValuesAs() matcher to compare LinkValue() types
  - we use semicolons to separate values in CSV and TSV representations;
    this is slightly iffy. The idea was copied over from the alpha.

Things not done:

  - I wanted to make FieldValue.isLink() and FieldValue.isList() into
    default methods, but there's a bug in OGNL (used by thymeleaf) which
    hides default methods from the templates.  This is fixed in OGNL
    3.0.12 (we're currently on 3.0.8).

om-sharma commented on src/main/java/uk/gov/register/presentation/Cardinality.java:
```diff
@@ -0,0 +1,37 @@
+     */
+    @JsonCreator
+    public static Cardinality fromId(String id) {
+        return cardinalityMap.get(id);
```
we can change the implementation as `Stream.of(Cardinality.values()).filter(c -> c.getId().equals(id)).findFirst().get();`,  which will remove the cardinalityMap presence.

philandstuff replied:
I'm ambivalent about this. Happy to change it if you prefer it but I'm not bothered.

om-sharma replied:
I am more concerned about duplicating all enum objects in separate map, which is already provided by enums in some form.

om-sharma commented:
Can you please rebase your branch with master. I can not locally see that the fields are rendered properly as a list. This is because of the 'registers.yaml' file is outdated here which is now changed in master.

philandstuff commented:
rebased.

philandstuff committed:
remove duplication of Cardinality values

The cardinalityMap construction repeated the list of all Cardinality
values, which is brittle if we ever add other cardinalities (say if we
explicitly state which fields are optional?)

This replaces the implementation with one that uses the values()
function to dynamically inspect every enum value.
